README.md
# prestashop-privat24
Privat24 payment module for Prestashop 1.6


CHANGELOG.txt
0.1
Initial version.

0.1.1
Try to upload module to PrestaShop market.

0.2
Check order currency and paid amount before changing order state.

0.2.1 - 15 May 2017
Improve module settings form. Translations updated.


translations/ru.php
<?php

global $_MODULE;
$_MODULE = array();
$_MODULE['<{privat24}default-bootstrap>privat24_23ca197d0452e82b973212819d663e32'] = 'Приват 24';
$_MODULE['<{privat24}default-bootstrap>privat24_b66cffdd06c6e18190a96632f50f9b05'] = 'Предоставляет платежный шлюз для Приват 24';
$_MODULE['<{privat24}default-bootstrap>privat24_44d7b2036166452a10240ddf1d7e93b0'] = 'Внимание! Все настройки будут удалены. Продолжить?';
$_MODULE['<{privat24}default-bootstrap>privat24_d712c500f00a67b86757e49eeff07312'] = 'Пожалуйста, настройте аккаунт мерчанта.';
$_MODULE['<{privat24}default-bootstrap>privat24_2e70d995f0032b8859dd5ef37162087f'] = 'ID мерчанта обязателен.';
$_MODULE['<{privat24}default-bootstrap>privat24_9c902c9eae767a6cd2cea030ad52f7d4'] = 'ID мерчанта должен быть числом.';
$_MODULE['<{privat24}default-bootstrap>privat24_c4d12d07b81b2d8ce4a0c4bbe1a8bb59'] = 'Пароль мерчанта обязателен.';
$_MODULE['<{privat24}default-bootstrap>privat24_643e540214f8cd92da56e922122392d4'] = 'Пожалуйста, проверьте что следующие Email адреса введены правильно:';
$_MODULE['<{privat24}default-bootstrap>privat24_19d88b422f8d77143357b977a9881cf8'] = 'При включении уведомления введите один или несколько адресов.';
$_MODULE['<{privat24}default-bootstrap>privat24_9f3a9b8720c8566caee5adffb9ca5bbe'] = 'Настройки были успешно сохранены.';
$_MODULE['<{privat24}default-bootstrap>privat24_136d8768eaf31f6d502cf402187974b5'] = 'Настройки Приват 24';
$_MODULE['<{privat24}default-bootstrap>privat24_229a7ec501323b94db7ff3157a7623c9'] = 'ID мерчанта';
$_MODULE['<{privat24}default-bootstrap>privat24_315f3c78ea54c8dadb082acc328dd5bd'] = 'Пароль мерчанта';
$_MODULE['<{privat24}default-bootstrap>privat24_7f77e3d7178f0b1979d13c101a456450'] = 'Режим отладки';
$_MODULE['<{privat24}default-bootstrap>privat24_8ea770fd89ab2a86fd9a08f24e3d846c'] = 'Записывать все ответы от Приват24 в каталоге';
$_MODULE['<{privat24}default-bootstrap>privat24_a3418ce40785563cd409d2dcaff789cf'] = 'Отправлять емайл когда платеж принят';
$_MODULE['<{privat24}default-bootstrap>privat24_9790b70b0772c3b35108155cf340f315'] = 'Емайлы';
$_MODULE['<{privat24}default-bootstrap>privat24_12b955eebbb7365c94195042edf06427'] = 'Емайлы можно разделять запятыми и переводами строки.';
$_MODULE['<{privat24}default-bootstrap>privat24_c9cc8cce247e49bae79f15173ce97354'] = 'Сохранить';
$_MODULE['<{privat24}default-bootstrap>privat24_b6b287fcde753d63583c953ff29f3d17'] = 'Прием платежей Приват24';
$_MODULE['<{privat24}default-bootstrap>payment_execution_d300f22155b4654603d6a19f184e584e'] = 'В течении нескольких секунд Вы будете перенаправлены на сайт Приват 24...';


translations/uk.php
<?php

global $_MODULE;
$_MODULE = array();
$_MODULE['<{privat24}default-bootstrap>privat24_23ca197d0452e82b973212819d663e32'] = 'Приват24';
$_MODULE['<{privat24}default-bootstrap>privat24_b66cffdd06c6e18190a96632f50f9b05'] = 'Дозволяє використовувати платіжну систему Приват24';
$_MODULE['<{privat24}default-bootstrap>privat24_44d7b2036166452a10240ddf1d7e93b0'] = 'Бажаєте видалити цей модуль ?';
$_MODULE['<{privat24}default-bootstrap>privat24_d712c500f00a67b86757e49eeff07312'] = 'Будь ласка, налаштуйте свій мерчант аккаунт.';
$_MODULE['<{privat24}default-bootstrap>privat24_2e70d995f0032b8859dd5ef37162087f'] = 'Потрібен ID мерчанта.';
$_MODULE['<{privat24}default-bootstrap>privat24_9c902c9eae767a6cd2cea030ad52f7d4'] = 'ID мерчанта може бути тільки числом.';
$_MODULE['<{privat24}default-bootstrap>privat24_c4d12d07b81b2d8ce4a0c4bbe1a8bb59'] = 'Потрібен пароль мерчанта.';
$_MODULE['<{privat24}default-bootstrap>privat24_643e540214f8cd92da56e922122392d4'] = 'Будь ласка, перевірьте наступні Email адреси:';
$_MODULE['<{privat24}default-bootstrap>privat24_19d88b422f8d77143357b977a9881cf8'] = 'Коли увімкнуто повідомлення про платіж потрібно ввести Email адреси.';
$_MODULE['<{privat24}default-bootstrap>privat24_9f3a9b8720c8566caee5adffb9ca5bbe'] = 'Налаштування були збережені.';
$_MODULE['<{privat24}default-bootstrap>privat24_136d8768eaf31f6d502cf402187974b5'] = 'Налаштування Приват24';
$_MODULE['<{privat24}default-bootstrap>privat24_229a7ec501323b94db7ff3157a7623c9'] = 'ID мерчанта';
$_MODULE['<{privat24}default-bootstrap>privat24_315f3c78ea54c8dadb082acc328dd5bd'] = 'Пароль мерчанта';
$_MODULE['<{privat24}default-bootstrap>privat24_7f77e3d7178f0b1979d13c101a456450'] = 'Вмикнути відлагодження';
$_MODULE['<{privat24}default-bootstrap>privat24_8ea770fd89ab2a86fd9a08f24e3d846c'] = 'Записувати відповіді від Приват24 API до каталогу';
$_MODULE['<{privat24}default-bootstrap>privat24_a3418ce40785563cd409d2dcaff789cf'] = 'Надіслати листа коли платіж буде прийнято';
$_MODULE['<{privat24}default-bootstrap>privat24_9790b70b0772c3b35108155cf340f315'] = 'Емайли';
$_MODULE['<{privat24}default-bootstrap>privat24_12b955eebbb7365c94195042edf06427'] = 'Емайл адреси ';
$_MODULE['<{privat24}default-bootstrap>privat24_c9cc8cce247e49bae79f15173ce97354'] = 'Зберегти';
$_MODULE['<{privat24}default-bootstrap>privat24_b6b287fcde753d63583c953ff29f3d17'] = 'Прийом платежів Приват24';


controllers/front/payment.php
<?php
/**
* 2016 Soft Industry
*
*   @author    Skorobogatko Alexei <a.skorobogatko@soft-industry.com>
*   @copyright 2016 Soft-Industry
*   @license   http://opensource.org/licenses/afl-3.0.php
*/

if (!defined('_PS_VERSION_')) {
    exit;
}

/**
 * Payment callback.
 *
 * @author skoro
 */
class Privat24PaymentModuleFrontController extends ModuleFrontController
{
    
    /**
     * @var boolean disable left sidebar.
     */
    public $display_column_left = false;
    
    /**
     * @see FrontController::initContent()
     */
    public function initContent()
    {
        parent::initContent();
        
        /** @var $cart Cart */
        /** @var $this->module Privat24 */

        if (!$this->module->active) {
            return;
        }
        
        $cart = $this->context->cart;
        $currency = new Currency((int)$cart->id_currency);
        $amount = $cart->getOrderTotal();
        
        $details = array();
        foreach ($cart->getProducts() as $product) {
            $details[] = $product['name'];
        }
        
        $customer = new Customer((int)$this->context->cart->id_customer);
        $this->module->validateOrder(
            $cart->id,
            Configuration::get('PRIVAT24_WAITINGPAYMENT_OS'),
            $amount,
            $this->module->displayName,
            null,
            array(),
            null,
            false,
            $customer->secure_key
        );
        
        $this->context->smarty->assign(array(
            'payment_url' => Privat24::PAY_PB_URL,
            'currency' => $currency->iso_code,
            'details' => implode(', ', $details),
            'amount' => $amount,
            'merchant_id' => $this->module->merchant_id,
            'order_id' => $cart->id,
            'return_url' => $this->context->link->getPageLink(
                'order-confirmation.php?key=' .
                $customer->secure_key .
                '&id_cart=' . (int) $this->context->cart->id .
                '&id_module=' . (int) $this->module->id .
                '&id_order=' . (int) $this->module->currentOrder
            ),
        ));
        
        $this->setTemplate('payment_execution.tpl');
    }
}


controllers/front/validation.php
<?php
/**
* 2016 Soft Industry
*
*   @author    Skorobogatko Alexei <a.skorobogatko@soft-industry.com>
*   @copyright 2016 Soft-Industry
*   @license   http://opensource.org/licenses/afl-3.0.php
*/

if (!defined('_PS_VERSION_')) {
    exit;
}

/**
 * Validate callback.
 *
 * @author skoro
 */
class Privat24ValidationModuleFrontController extends ModuleFrontController
{
    
    /**
     * @see FrontController::postProcess()
     */
    public function postProcess()
    {
        // Log requests from Privat API side in Debug mode.
        if (Configuration::get('PRIVAT24_DEBUG_MODE')) {
            $logger = new FileLogger();
            $logger->setFilename(_PS_ROOT_DIR_.'/log/'.$this->module->name.'_'.date('Ymd_His').'_response.log');
            $logger->logError($_POST);
        }
        
        $payment = array();
        parse_str(Tools::getValue('payment'), $payment);
        $hash = sha1(md5(Tools::getValue('payment') . $this->module->merchant_password));

        if ($payment && $hash === Tools::getValue('signature')) {
            if ($payment['state'] == 'ok') {
                $state = Configuration::get('PRIVAT24_WAITINGPAYMENT_OS');
                $cart_id = (int)$payment['order'];
                $order = new Order(Order::getOrderByCartId($cart_id));
                if (!Validate::isLoadedObject($order)) {
                    PrestaShopLogger::addLog('Privat24: cannot get order by cart id ' . $cart_id, 3);
                    die();
                }
                if ($order->getCurrentState() != $state) {
                    PrestaShopLogger::addLog(
                        sprintf(
                            'Privat24: order id %s current state %s !== expected state %s',
                            $order->id,
                            $order->getCurrentState(),
                            $state
                        ),
                        3
                    );
                    die();
                }
                
                // Check paid currency and paid amount.
                $id_currency = Currency::getIdByIsoCode($payment['ccy']);
                if (!$id_currency) {
                    PrestaShopLogger::addLog(
                        sprintf(
                            'Privat24: order id %s cannot get currency id by iso code: %s',
                            $order->id,
                            $payment['ccy']
                        ),
                        3
                    );
                    die();
                }
                if ($order->id_currency != $id_currency) {
                    PrestaShopLogger::addLog(
                        sprintf(
                            'Privat 24: order id %s, order currency id %s does not match with %s',
                            $order->id,
                            $order->id_currency,
                            $id_currency
                        ),
                        3
                    );
                    die();
                }
                if ((float)$order->total_paid != (float)$payment['amt']) {
                    PrestaShopLogger::addLog(
                        sprintf(
                            'Privat 24: order id %s order total paid %s does not match %s',
                            $order->id,
                            $order->total_paid,
                            $payment['amt']
                        ),
                        3
                    );
                    die();
                }
                
                $order_history = new OrderHistory();
                $order_history->id_order = $order->id;
                $order_history->changeIdOrderState(_PS_OS_PAYMENT_, $order->id);
                $order_history->addWithemail();
                $this->setPaymentTransaction($order, $payment);
                $this->module->paymentNotify($order, $payment);
                PrestaShopLogger::addLog(
                    sprintf(
                        'Privat24 payment accepted: order id: %s, amount: %s, ref: %s',
                        $order->id,
                        $payment['amt'],
                        $payment['ref']
                    ),
                    1
                );
            } else {
                PrestaShopLogger::addLog(
                    sprintf(
                        'Privat24 payment failed: state: %s, order: %s, ref: %s',
                        $payment['state'],
                        $payment['order'],
                        $payment['ref']
                    ),
                    3,
                    null,
                    null,
                    null,
                    true
                );
            }
        } else {
            PrestaShopLogger::addLog('Privat24: Payment callback bad signature.', 3, null, null, null, true);
        }
        
        die();
    }
    
    /**
     * Fill transaction_id in order payments.
     *
     * @param Order $order
     * @param array $payment payment data from gateway.
     */
    protected function setPaymentTransaction($order, array $payment)
    {
        Db::getInstance()->execute('
            UPDATE `'._DB_PREFIX_.'order_payment`
            SET transaction_id = "' . pSQL($payment['ref']) . '"
            WHERE order_reference = "' . pSQL($order->reference) . '"
                AND amount = ' . pSQL($payment['amt']) . '
                AND payment_method = "' . pSQL($this->module->displayName) . '"
                AND transaction_id = ""
        ');
    }
}


privat24.php
<?php
/**
* 2016 Soft Industry
*
*   @author    Skorobogatko Alexei <a.skorobogatko@soft-industry.com>
*   @copyright 2016 Soft-Industry
*   @license   http://opensource.org/licenses/afl-3.0.php
*/

if (!defined('_PS_VERSION_')) {
    exit;
}

/**
 * Privat24
 *
 * Provides payment gateway for Privat24 service.
 *
 * @author skoro
 */
class Privat24 extends PaymentModule
{
    
    /**
     * @var string
     */
    public $merchant_id;
    
    /**
     * @var string
     */
    public $merchant_password;
    
    /**
     * API url for payment card.
     */
    const PAY_PB_URL = 'https://api.privatbank.ua/p24api/ishop';
    
    /**
     * Contructor.
     */
    public function __construct()
    {
        $this->name = 'privat24';
        $this->version = '0.2.1';
        $this->author = 'Soft Industry';
        $this->tab = 'payments_gateways';
        $this->ps_versions_compliancy = array('min' => '1.6', 'max' => '1.6.99.99');
        $this->need_instance = 0;
        $this->bootstrap = true;
        $this->module_key = '98300a9bd7f3e615cc475c9e56cfef25';
        
        $this->merchant_id = Configuration::get('PRIVAT24_MERCHANT_ID');
        $this->merchant_password = Configuration::get('PRIVAT24_MERCHANT_PASSWORD');
        
        parent::__construct();
        
        $this->displayName = $this->l('Privat24');
        $this->description = $this->l('Provides Privat24 payment gateway.');
        
        $this->confirmUninstall = $this->l('Are you sure you want to uninstall ?');
        
        if (empty($this->merchant_id) || empty($this->merchant_password)) {
            $this->warning = $this->l('Please setup your Privat24 merchant account.');
        }
        
        $this->installMails();
    }
    
    /**
     * Module installation.
     */
    public function install()
    {
        return parent::install()
                && $this->registerHook('payment')
                && $this->registerHook('paymentReturn')
                && $this->createOrderState()
                && $this->installMails();
    }
    
    /**
     * Module uninstallation.
     */
    public function uninstall()
    {
        if (!parent::uninstall()) {
            return false;
        }
        
        Configuration::deleteByName('PRIVAT24_WAITINGPAYMENT_OS');
        Configuration::deleteByName('PRIVAT24_MERCHANT_ID');
        Configuration::deleteByName('PRIVAT24_MERCHANT_PASSWORD');
        Configuration::deleteByName('PRIVAT24_DEBUG_MODE');
        Configuration::deleteByName('PRIVAT24_PAYMENT_NOTIFY');
        Configuration::deleteByName('PRIVAT24_NOTIFY_EMAILS');
        
        $this->uninstallMails();
        
        return true;
    }
    
    /**
     * Create module order status.
     *
     * @return boolean
     */
    protected function createOrderState()
    {
        if (!Configuration::get('PRIVAT24_WAITINGPAYMENT_OS')) {
            $os = new OrderState();
            $os->name = array();
            foreach (Language::getLanguages(false) as $language) {
                switch (Tools::strtolower($language['iso_code'])) {
                    case 'ru':
                        $status = 'Ожидание оплаты Приват24';
                        break;
                    case 'ua':
                        $status = 'Очікування платежу Приват24';
                        break;
                    default:
                        $status = 'Waiting payment ' . $this->displayName;
                }
                $os->name[(int)$language['id_lang']] = $status;
            }
            $os->color = '#4169E1';
            $os->hidden = false;
            $os->send_email = false;
            $os->delivery = false;
            $os->logable = false;
            $os->invoice = false;
            if ($os->add()) {
                Configuration::updateValue('PRIVAT24_WAITINGPAYMENT_OS', $os->id);
                copy(dirname(__FILE__) . '/logo.gif', dirname(__FILE__) . '/../../img/os/' . (int)$os->id . '.gif');
            } else {
                return false;
            }
            return true;
        }
    }
    
    /**
     * Install mail templates.
     */
    protected function installMails()
    {
        $text = 'privat24_payment.txt';
        $html = 'privat24_payment.html';
        foreach (Language::getLanguages() as $language) {
            $dir = $this->local_path . 'mails/' . $language['iso_code'];
            if (file_exists($dir)) {
                $this->copyFile($dir . '/' . $text, _PS_MAIL_DIR_ . $language['iso_code'] . '/' . $text);
                $this->copyFile($dir . '/' . $html, _PS_MAIL_DIR_ . $language['iso_code'] . '/' . $html);
            }
        }
        
        return true;
    }
    
    /**
     * Uninstall mail templates.
     */
    protected function uninstallMails()
    {
        foreach (Language::getLanguages() as $language) {
            $dir = _PS_MAIL_DIR_ . $language['iso_code'];
            foreach (array('txt', 'html') as $ext) {
                $file = $dir . '/privat24_payment.' . $ext;
                if (file_exists($file)) {
                    @unlink($file);
                }
            }
        }
        return true;
    }
    
    /**
     * Copy file to destination.
     * @param string $source
     * @param string $dest
     */
    protected function copyFile($source, $dest)
    {
        if (@copy($source, $dest) == false) {
            $this->_errors[] = sprintf('Copy "%s" to "%s" failed.', $source, $dest);
        }
    }
    
    /**
     * Parse emails from a string data.
     *
     * @param string $data
     * @return array
     */
    protected function validateEmails($data)
    {
        $lines = explode("\n", $data);
        $lines = array_filter($lines, 'trim');
        $data = implode(',', $lines);
        $data = explode(',', trim($data));
        $data = array_filter($data);
        $emails = array();
        $errors = array();

        foreach ($data as $email) {
            $email = trim($email);
            if (Validate::isEmail($email)) {
                $emails[] = $email;
            } else {
                $errors[] = $email;
            }
        }

        return array(
            'emails' => $emails,
            'errors' => $errors,
        );
    }
    
    /**
     * Validate settings form POST data.
     *
     * @return array
     * @since 0.2.1
     */
    protected function postValidate()
    {
        $errors = array();
        
        if (Tools::isSubmit('submit' . $this->name)) {
            $merchant_id = Tools::getValue('PRIVAT24_MERCHANT_ID');
            if (!$merchant_id) {
                $errors[] = $this->l('Merchant ID is required.');
            }
            if (!Validate::isUnsignedInt($merchant_id)) {
                $errors[] = $this->l('Merchant ID must be a number.');
            }

            if (!Tools::getValue('PRIVAT24_MERCHANT_PASSWORD')) {
                $errors[] = $this->l('Merchant password is required.');
            }

            $emails = $this->validateEmails(Tools::getValue('PRIVAT24_NOTIFY_EMAILS'));
            if ($emails['errors']) {
                $errors[] = $this->l('Please check these emails are correct: ') .
                    implode(', ', $emails['errors']);
            }
            
            $notify = (bool) Tools::getValue('PRIVAT24_PAYMENT_NOTIFY');
            if ($notify && empty($emails['emails']) && empty($emails['errors'])) {
                $errors[] = $this->l('Email(s) needed when notify is enabled.');
            }
        }
        
        return $errors;
    }
    
    /**
     * Process module configuration form.
     *
     * @return string
     */
    public function getContent()
    {
        $output = $status = '';
        
        if (Tools::isSubmit('submit' . $this->name)) {
            $errors = $this->postValidate();

            if (empty($errors)) {
                $merchant_id = Tools::getValue('PRIVAT24_MERCHANT_ID');
                $merchant_password = Tools::getValue('PRIVAT24_MERCHANT_PASSWORD');
                $notify = (bool) Tools::getValue('PRIVAT24_PAYMENT_NOTIFY');
                $emails = Tools::getValue('PRIVAT24_NOTIFY_EMAILS');

                $backend_uri = str_replace('index.php', '', $_SERVER['PHP_SELF']);

                Configuration::updateValue('PRIVAT24_BACKEND_URI', $backend_uri);
                Configuration::updateValue('PRIVAT24_MERCHANT_ID', $merchant_id);
                Configuration::updateValue('PRIVAT24_MERCHANT_PASSWORD', $merchant_password);
                Configuration::updateValue('PRIVAT24_DEBUG_MODE', (bool)Tools::getValue('PRIVAT24_DEBUG_MODE'));
                Configuration::updateValue('PRIVAT24_PAYMENT_NOTIFY', $notify);
                Configuration::updateValue('PRIVAT24_NOTIFY_EMAILS', $emails);

                $this->merchant_id = $merchant_id;
                $this->merchant_password = $merchant_password;
                
                $status .= $this->displayConfirmation($this->l('Settings have been updated.'));
            } else {
                foreach ($errors as $error) {
                    $status .= $this->displayError($error);
                }
            }
        }

        return $output . $status . $this->displayForm();
    }
    
    /**
     * Module configuration form.
     *
     * @return string
     */
    public function displayForm()
    {
        $helper = new HelperForm();
        
        $fields_form = array();
        $fields_form['cfg']['form'] = array(
            'legend' => array(
                'title' => $this->l('Privat24 settings'),
            ),
            'input' => array(
                array(
                    'type' => 'text',
                    'label' => $this->l('Merchant ID'),
                    'name' => 'PRIVAT24_MERCHANT_ID',
                    'required' => true,
                ),
                array(
                    'type' => 'text',
                    'label' => $this->l('Merchant password'),
                    'name' => 'PRIVAT24_MERCHANT_PASSWORD',
                    'size' => 60,
                    'required' => true,
                ),
                array(
                    'type' => 'switch',
                    'label' => $this->l('Enable debug mode'),
                    'name' => 'PRIVAT24_DEBUG_MODE',
                    'desc' => $this->l('Log all Privat24 API responses in ')
                                . _PS_ROOT_DIR_.'/log/'
                                . $this->l(' directory.'),
                    'is_bool' => true,
                    'default_value' => 0,
                    'values' => array(
                        array('value' => 1),
                        array('value' => 0),
                    ),
                ),
                array(
                    'type' => 'switch',
                    'label' => $this->l('Email notify when payment accepted'),
                    'name' => 'PRIVAT24_PAYMENT_NOTIFY',
                    'is_bool' => true,
                    'default_value' => 0,
                    'values' => array(
                        array('value' => 1),
                        array('value' => 0),
                    ),
                ),
                array(
                    'type' => 'textarea',
                    'label' => $this->l('Emails'),
                    'name' => 'PRIVAT24_NOTIFY_EMAILS',
                    'desc' => $this->l('Emails can be separated by comma or/and newlines.'),
                ),
            ),
            'submit' => array(
                'title' => $this->l('Save'),
            ),
        );
        
        $helper->submit_action = 'submit' . $this->name;
        $helper->tpl_vars = array(
            'fields_value' => $this->getConfigFieldsValues(),
        );
        
        return $helper->generateForm($fields_form);
    }
    
    /**
     * Get form fields configuration values.
     *
     * @return array
     * @since 0.2.1
     */
    public function getConfigFieldsValues()
    {
        return array(
            'PRIVAT24_MERCHANT_ID' => Tools::getValue('PRIVAT24_MERCHANT_ID', Configuration::get('PRIVAT24_MERCHANT_ID')),
            'PRIVAT24_MERCHANT_PASSWORD' => Tools::getValue('PRIVAT24_MERCHANT_PASSWORD', Configuration::get('PRIVAT24_MERCHANT_PASSWORD')),
            'PRIVAT24_DEBUG_MODE' => Tools::getValue('PRIVAT24_DEBUG_MODE', Configuration::get('PRIVAT24_DEBUG_MODE')),
            'PRIVAT24_PAYMENT_NOTIFY' => Tools::getValue('PRIVAT24_PAYMENT_NOTIFY', Configuration::get('PRIVAT24_PAYMENT_NOTIFY')),
            'PRIVAT24_NOTIFY_EMAILS' => Tools::getValue('PRIVAT24_NOTIFY_EMAILS', Configuration::get('PRIVAT24_NOTIFY_EMAILS')),
        );
    }
    
    /**
     * Implements hookPayment.
     */
    public function hookPayment($params)
    {
        if (!$this->active) {
            return;
        }
        
        if (isset($this->context->controller)) {
            $this->context->controller->addCSS($this->_path . 'views/css/privat24.css');
        }
        
        $this->smarty->assign(array(
            'this_path' => $this->_path,
            'this_path_ssl' => Tools::getShopDomainSsl(true, true).__PS_BASE_URI__.'modules/'.$this->name.'/',
            'id_cart' => (int) $params['cart']->id,
        ));

        return $this->display(__FILE__, 'payment.tpl');
    }
    
    /**
     * Not used in this module.
     */
    public function hookPaymentReturn()
    {
    }
    
    /**
     * Notify emails about accepted payment.
     *
     * @param Order $order order instance
     * @param array $payment data from payment gateway
     * @return bool
     */
    public function paymentNotify(Order $order, array $payment)
    {
        if (!Configuration::get('PRIVAT24_PAYMENT_NOTIFY')) {
            return true;
        }
        if (!Validate::isLoadedObject($order)) {
            return false;
        }
        $emails = $this->validateEmails(Configuration::get('PRIVAT24_NOTIFY_EMAILS'));
        if (!empty($emails['emails'])) {
            $template = 'privat24_payment';
            $subject = $this->l('Payment accepted via Privat24');
            $context = Context::getContext();
            $data = array(
                '{order_name}' => $order->getUniqReference(),
                '{order_link}' =>
                        _PS_BASE_URL_ .
                        Configuration::get('PRIVAT24_BACKEND_URI') .
                        $context->link->getAdminLink('AdminOrders', false) . '&vieworder&id_order=' . (int)$order->id,
                '{amount}' => $payment['amt'] . ' ' . $context->currency->iso_code,
                '{payment_transaction}' => $payment['ref'],
            );
            foreach ($emails['emails'] as $email) {
                Mail::Send((int)$order->id_lang, $template, $subject, $data, $email);
            }
        }
        return true;
    }
}
